File: IV_Data.m
classdef IV_Data
   properties
      file_name
      data_vg
      data_vd
      data_ig
      data_id
      vg_range
      vd_range
      vg_number_of_each
      vd_number_of_each
      vg_step
      vd_step
      ngf
      W
      L
   end

   methods

       function obj = IV_Data(path_string)
           [filepath,name,ext] = fileparts(path_string);
           obj.file_name = name;
           if strcmp(ext, ".mat")
               mat_info = load(path_string);
               info_obj = mat_info.MeasurementIV;
               obj.ngf = info_obj.NGF;
               obj.W = info_obj.W;
               obj.L = info_obj.L;
               obj.vg_range = [info_obj.VGStart, info_obj.VGStop];
               obj.vg_step = info_obj.VGstep;
               obj.data_vg = info_obj.VGset;
               obj.data_vd = info_obj.VDset;
               obj.data_ig = info_obj.IGmeas;
               obj.data_id = info_obj.IDmeas;
           end
           
           if strcmp(ext, ".csv")
                [vg_string_data, data_strings] = extractFileInfo(path_string);
                [data_table, contains_vg] = formatToTable(data_strings);
                [vg_data, vg_start, vg_step] = parseVgString(vg_string_data);
                
                if ~contains_vg
                    data_table = addVgToTable(data_table, vg_data);
                end
                
                % Set Data Arrays
                obj.data_vg = data_table.Vg;

                if any(strcmp('Vd', data_table.Properties.VariableNames))
                    obj.data_vd = data_table.Vd;
                end

                if any(strcmp('Ig', data_table.Properties.VariableNames))
                    obj.data_ig = data_table.Ig;
                end

                if any(strcmp('Id', data_table.Properties.VariableNames))
                    obj.data_id = data_table.Id;
                end

                % Set other parameters
                obj.vg_step = vg_step;
                obj.vg_range = [vg_start, vg_data(length(vg_data))];

                [ngf, w] = extractDimensionsFromFileName(filepath);
                obj.ngf = ngf;
                obj.W = w;
           end
            
           % Set number of occurences of each Vg and Vd
           gc = groupcounts(obj.data_vg);
           if all(gc == gc(1))
               obj.vg_number_of_each = gc(1);
           end

           gc = groupcounts(obj.data_vd);
           if all(gc == gc(1))
               obj.vd_number_of_each = gc(1);
           end

           obj.vd_range = [max(obj.data_vd), min(obj.data_vd)];
           obj.vd_step = obj.data_vd(2) - obj.data_vd(1); % Should fix to read from file
           
       end

       function results_matrix = getDataFromVg(obj, values)
           results_matrix = cell(1, length(values));
           if (any(values > obj.vg_range(1)) || any(values < obj.vg_range(2)))
            error(" Invalid Values outside range %.2f to %.2f", obj.vg_range(2), obj.vg_range(1))
           end

           for i=1:length(values)
               loc = (obj.data_vg == values(i));
               if not(loc)
                   error("Invalid VG - Vg = %.2f not in measurements", values(i))
               end

               results = zeros(obj.vg_number_of_each, 3);
               results(:, 1) = obj.data_vd(loc);
               results(:, 2) = obj.data_id(loc);
               if ~isempty(obj.data_ig)
                   results(:, 3) = obj.data_ig(loc);
               else 
                   results(:, 3) = nan(obj.vg_number_of_each, 1);
               end

               results_matrix{1, i} = results;
           end

       end

       function results_matrix = getDataFromVd(obj, values)
           results_matrix = cell(1, length(values));
           if (any(values > obj.vd_range(1)) || any(values < obj.vd_range(2)))
            error(" Invalid Values outside range %.2f to %.2f", obj.vd_range(2), obj.vd_range(1))
           end

           for i=1:length(values)
               loc = (obj.data_vd == values(i));
               if not(loc)
                   error("Invalid VD - Vd = %.2f not in measurements", values(i))
               end

               results = zeros(obj.vd_number_of_each, 3);
               results(:, 1) = obj.data_vg(loc);
               results(:, 2) = obj.data_id(loc);
               if ~isempty(obj.data_ig)
                   results(:, 3) = obj.data_ig(loc);
               else 
                   results(:, 3) = nan(obj.vd_number_of_each, 1);
               end

               results_matrix{1, i} = results;
           end

       end
   end
end

File: IV_Data.py
from re import L
import numpy as np
import pandas as pd
import os
import pathlib
import scipy.io as sio
from FilesFromDirectory import FilesFromDirectory

def get_important_data_from_csv(data_path):
    vg_filter = "Analysis.Setup.Vector.Graph.SetupInfo" # in the second column
    data_filter_label = "DataName" # In the first column
    data_filter_vals = "DataValue" # In the first column

    vg_row = []
    data = []

    with open(data_path, 'r') as f:
        for line in f.readlines():
            if vg_filter in line:
                vg_row = [entry.strip() for entry in line.split(",")]
            if data_filter_label in line:
                data.append([entry.strip() for entry in line.split(",")])
            if data_filter_vals in line:
                data.append([entry.strip() for entry in line.split(",")])
    
    return vg_row, data

def calculate_v_num_elem(step, v_range):
    if step < 0: 
        return len(np.arange(v_range[0], v_range[1] , step))
    else:
        return len(np.arange(v_range[1], v_range[0] , step))

def parse_sep_vg_entry(vg_values, vd_num_ea):
    tab_entry_start = vg_values[2].split('\t')
    tab_entry_end = vg_values[len(vg_values) - 1].split('\t')

    vg_data = [float(tab_entry_start[1])] * vd_num_ea
    
    for entry in vg_values[3:-1]:
        vg_data.extend([float(entry)] * vd_num_ea)
    
    vg_data.extend([float(tab_entry_end[0])] * vd_num_ea)
    
    return vg_data

class IV_Data:
    file_name = ""
    data_vg = []
    data_vd = []
    data_ig = []
    data_id = []
    vg_range = []
    vd_range = []
    vg_number_of_each = np.NaN
    vd_number_of_each = np.NaN
    vg_step = np.NaN
    vd_step = np.NaN
    ngf = np.NaN
    W = np.NaN
    L = np.NaN

    def __init__(self, file_path):
        if (os.path.isfile(file_path)):
            path = pathlib.Path(file_path)
            self.file_name = path.name
            extension = path.suffix

            if ".csv" in extension:
                vg, data = get_important_data_from_csv(path)
                df = pd.DataFrame(data[1:], columns = data[0])
                df = df.replace(r'', np.nan, regex=True)

                numeric_columns = [i for i in df.columns if i != 'DataName']
                for column in numeric_columns:
                    df[column] = df[column].astype(float)

                if ("Vd" in df.columns):
                    vd_values = df["Vd"].unique()
                    self.data_vd = df["Vd"].tolist()
                    self.vd_range = [vd_values.max(), vd_values.min()]
                    self.vd_step = vd_values[1] - vd_values[0]               
                    self.vd_number_of_each = calculate_v_num_elem(self.vd_step, self.vd_range)
                    vd_max_indices = df[df["Vd"] == vd_values.max()].index

                if not ("Vg" in df.columns):
                    index_vd_max = vd_max_indices.sort_values()
                    df["Vg"] = parse_sep_vg_entry(vg, index_vd_max[1] - index_vd_max[0])

                vg_values = df["Vg"].unique()
                self.data_vg = df["Vg"].tolist()
                self.vg_range = [max(vg_values), min(vg_values)]
                self.vg_step = vg_values[1] - vg_values[0]
                self.vg_number_of_each = calculate_v_num_elem(self.vg_step, self.vg_range)

                if ("Id" in df.columns):
                    self.data_id = df["Id"].tolist()

                if ("Ig" in df.columns):
                    self.data_ig = df["Ig"].tolist()


            elif ".mat" in extension:
                print("MAT")
                mat_contents = sio.loadmat(path)
                mat_struct = mat_contents["MeasurementIV"][0,0]
                self.ngf = mat_struct["NGF"][0,0]
                self.W = mat_struct["W"][0,0]
                self.L = mat_struct["L"][0,0]
                self.vg_step = mat_struct["VGstep"][0,0]
                self.vd_step = mat_struct["VDstep"][0,0]
                self.vd_step = mat_struct["VDstep"][0,0]
                self.vg_range = [mat_struct["VGStart"][0,0], mat_struct["VGStop"][0,0]]
                self.vd_range = [mat_struct["VDStart"][0,0], mat_struct["VDStop"][0,0]]
                self.data_vg = np.array(mat_struct["VGset"]).squeeze()
                self.data_vd = np.array(mat_struct["VDset"]).squeeze()
                self.data_ig = np.array(mat_struct["IGmeas"]).squeeze()
                self.data_id = np.array(mat_struct["IDmeas"]).squeeze()
                self.vg_number_of_each = np.count_nonzero(self.data_vg == self.vg_range[1])
                self.vd_number_of_each = np.count_nonzero(self.data_vd == self.vd_range[1])
            
            else:
                print("Not Valid Extension - must be csv or mat")

    def print_important_param(self):
        print(f"vg_step: {self.vg_step}, vg_range: {self.vg_range}, vd_step: {self.vd_step}, vd_range: {self.vd_range}, num_vg: {len(self.data_vg)}, num_vd: {len(self.data_vd)}")
        print(f"Number of Points: VG: {self.vg_number_of_each}  VD:{self.vd_number_of_each}")


dir_str = r"..\\..\\..\\OneDrive\\Documents\\University\\Fourth Year\\4A\\URA\\IV Raw Data"
files_obj = FilesFromDirectory(dir_str, "csv")


x = IV_Data(files_obj.relative_filepaths[6])
print(x.file_name)
x.print_important_param()
